loader/loader_windows.c

```diff
@@ -793,6 +793,133 @@ VkResult windows_read_data_files_in_registry(const struct loader_instance *inst,
+void sort_physical_devices_with_same_luid(struct loader_instance *inst, uint32_t sorted_devices_count,
+                                          struct loader_phys_dev_per_icd *sorted_array) {
+    bool app_is_vulkan_1_1 = loader_check_version_meets_required(LOADER_VERSION_1_1_0, inst->app_api_version);
+
+    for (uint32_t i = 0; sorted_devices_count > 1 && i < sorted_devices_count - 1; i++) {
+        for (uint32_t j = i + 1; j < sorted_devices_count; j++) {
+            // Only want to reorder physical devices if their ICD's LUID's match
+            if ((sorted_array[i].windows_adapter_luid.HighPart != sorted_array[j].windows_adapter_luid.HighPart) &&
+                (sorted_array[i].windows_adapter_luid.LowPart != sorted_array[j].windows_adapter_luid.LowPart)) {
+                continue;
+            }
+
+            VkLayeredDriverUnderlyingApiMSFT underlyingAPI = VK_LAYERED_DRIVER_UNDERLYING_API_NONE_MSFT;
+            VkPhysicalDeviceLayeredDriverPropertiesMSFT layered_driver_properties_msft = {0};
+            layered_driver_properties_msft.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_LAYERED_DRIVER_PROPERTIES_MSFT;
+            VkPhysicalDeviceProperties2 props2 = {0};
+            props2.sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_PROPERTIES_2;
+            props2.pNext = (void *)&layered_driver_properties_msft;
+
+            // Because there may be multiple physical devices associated with each ICD, we need to check each physical device
+            // whether it is layered
+            for (uint32_t k = 0; k < sorted_array[i].device_count; k++) {
+                VkPhysicalDeviceProperties dev_props = {0};
+                sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties(sorted_array[i].physical_devices[k], &dev_props);
+
+                bool device_is_1_1_capable =
+                    loader_check_version_meets_required(LOADER_VERSION_1_1_0, loader_make_version(dev_props.apiVersion));
+
+                PFN_vkGetPhysicalDeviceProperties2 GetPhysDevProps2 = NULL;
+                if (app_is_vulkan_1_1 && device_is_1_1_capable) {
+                    GetPhysDevProps2 = sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties2;
+                } else {
+                    GetPhysDevProps2 =
+                        (PFN_vkGetPhysicalDeviceProperties2)sorted_array[i].icd_term->dispatch.GetPhysicalDeviceProperties2KHR;
+                }
+                if (GetPhysDevProps2) {
+                    GetPhysDevProps2(sorted_array[i].physical_devices[k], &props2);
+                    if (layered_driver_properties_msft.underlyingAPI != VK_LAYERED_DRIVER_UNDERLYING_API_NONE_MSFT) {
+                        underlyingAPI = layered_driver_properties_msft.underlyingAPI;
+                        break;
+                    }
+                }
+            }
+            if (underlyingAPI == VK_LAYERED_DRIVER_UNDERLYING_API_D3D12_MSFT) {
```

Comment:
Just to make sure I'm reading this correctly, this function is responsible for sorting entire ICDs, not individual devices, right? So, given the scenario with the following ICD/device pairings:
* ICD A
   * GPU 0
* ICD B
   * GPU 0 (layered)
   * GPU 1 (layered)

If the DXGI sort order says that GPU 0 should come first, and then GPU 1, the final Vulkan order would be:
* GPU 0
* GPU 0 (layered)
* GPU 1 (layered)

However, if GPU 1 is the one that's supposed to be first, won't this produce the incorrect sorting:
* GPU 0
* GPU 1 (layered)
* GPU 0 (layered)

Since the layered *ICD* ends up sorted after the native ICD? Or is there a second-level sort that I'm not seeing? I'll try to experimentally confirm this, but wanted to clarify what your intended behavior is here.

Reply:
In the example you provided, there should not be any sorting to occur because both GPU's are layered. The swapping of ICD's only occurs if they correspond to the same LUID, one is layered while the other is not, and the layered one is earlier in the list. But please do experiment around with it. 

I may take your experiments and create test cases from it to solidify the desired behavior in different circumstances.

Reply:
Now that I've stepped through this, I think this is correct. The naming was a little confusing, I'd thought that the sorting was changing the order of whole ICDs, but now I see that `loader_phys_dev_per_icd` is just a physical device with a *reference* to which ICD it came from. So, only the physical devices should be sorted. From playing around a bit, things seem to be working nicely with my one bugfix implemented.

Reply:
Hopefully the rename to loader_icd_physical_devices is less confusing. I couldn't come up with a better name, but per_icd gives a very different impression of what it represents.